### ISSUE TITLE
Assertion hit due to daemon-reload imposed race

### ISSUE THREAD

**Issue by jnpkrn** [systemd/systemd#10627]
**systemd version the issue has been seen with**
> 239 (`systemd-239-6.git9f3aed1.fc30.x86_64`)

**Used distribution**
 > Fedora Rawhide

**Steps to reproduce the problem**

```
# > /run/systemd/system/_test.service cat <<EOF
[Service]
Type=simple
WorkingDirectory=/tmp
ExecStart=/usr/bin/sleep 100 
EOF

# while :; do \
  dbus-send --system --print-reply --dest=org.freedesktop.systemd1 \
  /org/freedesktop/systemd1 org.freedesktop.systemd1.Manager.Reload; \
  sleep 1; done & { systemctl edit --full _test.service; fg; }
```

When editor opens, do at least a single change, e.g. add a new empty
line.

Result immediately observable:

```
method return time=1541195573.509673 sender=:1.1 -> destination=:1.129 serial=21465 reply_serial=2
method return time=1541195574.632637 sender=:1.1 -> destination=:1.130 serial=22346 reply_serial=2
method return time=1541195575.761765 sender=:1.1 -> destination=:1.131 serial=23231 reply_serial=2
[...]
Error org.freedesktop.DBus.Error.NoReply: Message recipient disconnected from message bus without replying
Error org.freedesktop.DBus.Error.TimedOut: Failed to activate service 'org.freedesktop.systemd1': timed out (service_start_timeout=25000ms)
Error org.freedesktop.DBus.Error.TimedOut: Failed to activate service 'org.freedesktop.systemd1': timed out (service_start_timeout=25000ms)
Error org.freedesktop.DBus.Error.TimedOut: Failed to activate service 'org.freedesktop.systemd1': timed out (service_start_timeout=25000ms)
```

Actual result as captured in journal:

```
Nov 02 22:52:55 systemd[1]: Reloading.
Nov 02 22:52:56 systemd[1]: Reloading.
Nov 02 22:53:06 systemd[1]: Assertion '!m->queued_message' failed at ../src/core/dbus-manager.c:1332, function method_reload(). Aborting.
Nov 02 22:53:06 audit[2341]: ANOM_ABEND auid=4294967295 uid=0 gid=0 ses=4294967295 subj=system_u:system_r:init_t:s0 pid=2341 comm="systemd" exe="/usr/lib/systemd/systemd" sig=6 res=1
Nov 02 22:53:06 systemd-coredump[2342]: Due to PID 1 having crashed coredump collection will now be turned off.
Nov 02 22:53:06 systemd[1]: Caught <ABRT>, dumped core as pid 2341.
[...]
Nov 02 22:53:06 systemd[1]: Freezing execution.
[...]
Nov 02 22:53:07 systemd-coredump[2342]: Process 2341 (systemd) of user 0 dumped core.
                                        
                                        Stack trace of thread 2341:
                                        #0  0x00007f3b94e4284b kill (libc.so.6)
                                        #1  0x00005620bbde6eda crash (systemd)
                                        #2  0x00007f3b94fe3070 __restore_rt (libpthread.so.0)
                                        #3  0x00007f3b94e4254f raise (libc.so.6)
                                        #4  0x00007f3b94e2c895 abort (libc.so.6)
                                        #5  0x00007f3b951c1d5f log_assert_failed_realm (libsystemd-shared-239.so)
                                        #6  0x00005620bbded5b1 method_reload (systemd)
                                        #7  0x00007f3b95291d2b method_callbacks_run (libsystemd-shared-239.so)
                                        #8  0x00007f3b95293316 bus_process_object (libsystemd-shared-239.so)
                                        #9  0x00007f3b952a1d5f process_message (libsystemd-shared-239.so)
                                        #10 0x00007f3b952a340c io_callback (libsystemd-shared-239.so)
                                        #11 0x00007f3b952c9ef0 source_dispatch (libsystemd-shared-239.so)
                                        #12 0x00007f3b952cbc55 sd_event_dispatch (libsystemd-shared-239.so)
                                        #13 0x00007f3b952cbde0 sd_event_run (libsystemd-shared-239.so)
                                        #14 0x00005620bbe2815c manager_loop (systemd)
                                        #15 0x00005620bbde278b invoke_main_loop (systemd)
                                        #16 0x00007f3b94e2e413 __libc_start_main (libc.so.6)
                                        #17 0x00005620bbde441e _start (systemd)

```

Actual backtrace:

```
#0  kill

#1  crash
    (sig=6)
    at ../src/core/main.c:183

#2  <signal handler called>

#3  raise

#4  abort

#5  log_assert_failed_realm
    (realm=realm@entry=LOG_REALM_SYSTEMD, text=text@entry=0x5620bbe8f2de
    "!m->queued_message", file=file@entry=0x5620bbe8e610
    "../src/core/dbus-manager.c", line=line@entry=1332,
    func=func@entry=0x5620bbe8faf8 <__PRETTY_FUNCTION__.18462>
    "method_reload")
    at ../src/basic/log.c:806

#6  method_reload
    (message=0x5620bcb211e0, userdata=0x5620bca79620, error=0x7fffa7f4b850)
    at ../src/core/dbus-manager.c:133

#7  method_callbacks_run
    (found_object=0x7fffa7f4b937, require_fallback=false,
    c=0x5620bcbf3dd0, m=0x5620bcb211e0, bus=0x5620bcc38120)
    at ../src/libsystemd/sd-bus/bus-objects.c:404

#8  object_find_and_run
    (bus=0x5620bcc38120, m=0x5620bcb211e0, p=0x5620bccd11d8
    "/org/freedesktop/systemd1", require_fallback=false,
    found_object=0x7fffa7f4b937)
     at ../src/libsystemd/sd-bus/bus-objects.c:1262

#9  bus_process_object
    (bus=bus@entry=0x5620bcc38120, m=m@entry=0x5620bcb211e0)
    at ../src/libsystemd/sd-bus/bus-objects.c:1378

#10 process_message
    (m=0x5620bcb211e0, bus=0x5620bcc38120)
    at ../src/libsystemd/sd-bus/sd-bus.c:2663

#11 process_running
    (ret=0x0, priority=<optimized out>, hint_priority=<optimized out>,
    bus=0x5620bcc38120)
    at ../src/libsystemd/sd-bus/sd-bus.c:2705

#12 bus_process_internal
    (bus=0x5620bcc38120, hint_priority=<optimized out>,
    priority=<optimized out>, ret=0x0)
    at ../src/libsystemd/sd-bus/sd-bus.c:2924

#13 io_callback
    (s=<optimized out>, fd=<optimized out>, revents=<optimized out>,
    userdata=0x5620bcc38120)
    at ../src/libsystemd/sd-bus/sd-bus.c:3304

#14 source_dispatch
    (s=s@entry=0x5620bcc38dc0)
    at ../src/libsystemd/sd-event/sd-event.c:3103

#15 sd_event_dispatch
    (e=0x5620bca79bb0)
    at ../src/libsystemd/sd-event/sd-event.c:3516

#16 sd_event_run
    (e=0x5620bca79bb0, timeout=18446744073709551615)
    at ../src/libsystemd/sd-event/sd-event.c:3573

#17 manager_loop
    (m=m@entry=0x5620bca79620)
    at ../src/core/manager.c:2855

#18 invoke_main_loop
    (ret_error_message=0x7fffa7f4bcd8, ret_switch_root_init=<synthetic
    pointer>, ret_switch_root_dir=<synthetic pointer>,
    ret_fds=0x7fffa7f4bce8, ret_shutdown_verb=<synthetic pointer>,
    ret_retval=<synthetic pointer>, ret_reexecute=<synthetic pointer>,
    m=0x5620bca79620)
    at ../src/core/main.c:1630

#19 main
    (argc=<optimized out>, argv=0x7fffa7f4bfb8)
    at ../src/core/main.c:2417
```

Needless to say, many things not working (even at surprising places, like `su`),
forced reboot imminent.

The problem doesn't exhibit when `systemctl edit` is run with `--no-reload`.

EDIT: doesn't appear reproducible with v219 on other distro I have handy.

**Comment by jnpkrn** [comment]
Btw. I hit something vaguely similar years ago:
https://bugzilla.redhat.com/show_bug.cgi?id=1244342

But that appeared to be so accidental and complex to reproduce
I didn't bother attempting so.

Another possible hit whereby daemon-reexec (solely or in addition
to daemon-reload may be the interfering point, note that there
was no systemd update in dnf transaction with the former):
https://bugzilla.redhat.com/show_bug.cgi?id=1618954

Both appear to be closer to each other than to this very observation,
though.

**Comment by fbuihuu** [comment]
Probably introduced by e0a085811de1d1976c019afcfc2e4e74f590cc9f

**Comment by poettering** [comment]
@fbuihuu ah, thanks for the hint, I think you are right.

**Comment by poettering** [comment]
Fix pending in #10763
